cpp/src/io/csv/writer_impl.cu

```diff
@@ -0,0 +1,497 @@
+ * @param options ...
+ * @param mr...
+ * @return strings_column_view instance formated for CSV column output.
+**/
+// strings_column_view column_to_strings_csv(column_view const& column,
+//                                           writer_options const& options,
+//                                           rmm::mr::device_memory_resource* mr = nullptr) {
+//   //TODO;
+//   //
+//   column_to_strings_fn col2str{options, mr};
+//   auto ret = col2str.template operator()<bool>(column); // check instantiation: okay
+  
+//   return strings_column_view{column}; // for now
+// }
+
+} // unnamed namespace
+
+// Forward to implementation
+writer::writer(std::unique_ptr<data_sink> sink,
+               writer_options const& options,
+               rmm::mr::device_memory_resource* mr)
+  : _impl(std::make_unique<impl>(std::move(sink), options, mr))
+{
+}
+
+// Destructor within this translation unit
+writer::~writer() = default;
+
+
+writer::impl::impl(std::unique_ptr<data_sink> sink,
+                   writer_options const &options,
+                   rmm::mr::device_memory_resource *mr):
+  out_sink_(std::move(sink)),
+  mr_(mr),
+  options_(options)
+{
+}
+
+// write the header: column names:
+//
+void writer::impl::write_chunked_begin(table_view const& table,
+                                       const table_metadata *metadata,
+                                       cudaStream_t stream)
+{
+  if ( options_.include_header() )
+    {
+      CUDF_EXPECTS( metadata != nullptr, "Unexpected null metadata."); 
+      CUDF_EXPECTS( metadata->column_names.size() == static_cast<size_t>(table.num_columns()),
+                    "Mismatch between column headers and table columns.");
+
+      std::string delimiter_str{options_.inter_column_delimiter()};
+
+      //avoid delimiter after last element:
+      //
+      std::stringstream ss;
+      std::copy(metadata->column_names.begin(), metadata->column_names.end()-1,
+                std::ostream_iterator<std::string>(ss, delimiter_str.c_str()));
+      ss<<metadata->column_names.back();
+
+      out_sink_->host_write(ss.str().data(), ss.str().size());
+    }
+}
+
+
+void writer::impl::write_chunked(strings_column_view const& strings_column,
+                                 const table_metadata *metadata,
+                                 cudaStream_t stream)
+{
+  //algorithm outline:
+  //
+  //  for_each(strings_column.begin(), strings_column.end(),
+  //           [sink = out_sink_](auto str_row) mutable {
+  //               auto host_buffer = str_row.host_buffer();
+  //               sink->host_write(host_buffer_.data(), host_buffer_.size());
+  //           });//or...sink->device_write(device_buffer,...);
+
+  auto pair_buff_offsets = cudf::strings::create_offsets(strings_column, stream, mr_);
+
+  auto num_rows = strings_column.size();
+  decltype(num_rows) num_offsets = pair_buff_offsets.second.size();
+
+  /* cannot use with CUDF_EXPECTS() because it's a macro...
```

Comment:
This is a confusing comment. Is this old code to be removed?

Reply:
It's just an attempt to pass a runtime message string to `CUDF_EXPECTS`. But that doesn't compile because `CUDF_EXPECTS` is a macro. Which begs the question why isn't it an inlined function, instead? I was hoping I'd find a workaround (perhaps by paranthesize-ing `reason` in  `CUDF_EXPECTS` macro; but that doesn't work, either).

So, I left the comment around perhaps someone has a better idea. Or, I can make the case to replace the macro with an inlined function.

Reply:
AFAIK, CUDF_EXPECTS is a macro because of __LINE__ and __FILE__ use.
I think the current assertion below is fine here, don't think it's worth expanding the error message for this.

Reply:
Well, you can pass `__FILE__` and `__LINE__` as function arguments at calling point: 
```
void dump_file_line(char const* const sfile, int line, std::string const& more)
{
    std::cout<<"File:line: "<<sfile<<":"<<line<<more<<'\n';
}
...
std::string str;
...
 dump_file_line(__FILE__, __LINE__, str);//this file...and this line
```

Reply:
>Well, you can pass __FILE__ and __LINE__ as function arguments at calling point

You can, it just makes the calling site a bit more verbose. I don't know if this is why the person who implemented CUDF_EXPECTS made it a macro, just guessing. You can definitely follow up on this. I commented here to ask for the commented out code to be removed if it's unused.

Reply:
Done.